src/python/pants/core/goals/deploy.py

```diff
@@ -0,0 +1,203 @@
+    TargetRootsToFieldSetsRequest,
+)
+from pants.engine.unions import union
+from pants.option.option_types import ArgsListOption
+
+logger = logging.getLogger(__name__)
+
+
+@union
+@dataclass(frozen=True)
+class DeployFieldSet(FieldSet, metaclass=ABCMeta):
+    """The FieldSet type for the `deploy` goal."""
+
+
+@dataclass(frozen=True)
+class DeployProcess:
+    name: str
+    publish_dependencies: tuple[Target, ...]
+    process: InteractiveProcess | None
+    description: str | None = None
+
+
+class DeploySubsystem(GoalSubsystem):
+    name = "experimental-deploy"
+    help = "Perform a deployment process."
+
+    required_union_implementation = (DeployFieldSet,)
+
+    args = ArgsListOption(
+        example="--force",
+        passthrough=True,
+        extra_help="Arguments to pass to the underlying tool.",
+    )
+
+
+@dataclass(frozen=True)
+class Deploy(Goal):
+    subsystem_cls = DeploySubsystem
+
+
+@rule_helper
+async def _find_publish_processes(targets: Iterable[Target]) -> PublishProcesses:
+    package_field_sets, publish_field_sets = await MultiGet(
+        Get(FieldSetsPerTarget, FieldSetsPerTargetRequest(PackageFieldSet, targets)),
+        Get(FieldSetsPerTarget, FieldSetsPerTargetRequest(PublishFieldSet, targets)),
+    )
+
+    return await Get(
+        PublishProcesses,
+        PublishProcessesRequest(
+            package_field_sets=package_field_sets.field_sets,
+            publish_field_sets=publish_field_sets.field_sets,
+        ),
+    )
+
+
+@rule_helper
+async def _invoke_process(
+    console: Console,
+    process: InteractiveProcess | None,
+    *,
+    names: Iterable[str],
+    success_status: str,
+    description: str | None = None,
+) -> tuple[int, tuple[str, ...]]:
+    results = []
+
+    if not process:
+        sigil = console.sigil_skipped()
+        status = "skipped"
+        if description:
+            status += f" {description}"
+        for name in names:
+            results.append(f"{sigil} {name} {status}.")
+        return 0, tuple(results)
+
+    logger.debug(f"Execute {process}")
+    res = await Effect(InteractiveProcessResult, InteractiveProcess, process)
+    if res.exit_code == 0:
+        sigil = console.sigil_succeeded()
+        status = success_status
+        prep = "to"
+    else:
+        sigil = console.sigil_failed()
+        status = "failed"
+        prep = "for"
+
+    if description:
+        status += f" {prep} {description}"
+
+    for name in names:
+        results.append(f"{sigil} {name} {status}")
+
+    return res.exit_code, tuple(results)
+
+
+@goal_rule
+async def run_deploy(console: Console, deploy_subsystem: DeploySubsystem) -> Deploy:
+    target_roots_to_deploy_field_sets = await Get(
+        TargetRootsToFieldSets,
+        TargetRootsToFieldSetsRequest(
+            DeployFieldSet,
+            goal_description=f"the `{deploy_subsystem.name}` goal",
+            no_applicable_targets_behavior=NoApplicableTargetsBehavior.error,
```

Comment:
Decided to error here instead of giving a warning because, being a side-efficting goal that is quite likely to be run in CI/CD pipelines, I would like to see those pipelines erroring when there are not targets to deploy, as it would highlight the fact that the pipeline has been wrongly configured.